titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/DefaultKubeJobManagementReconciler.java

```diff
@@ -0,0 +1,286 @@
+            return;
+        }
+
+        List<V1Node> nodes = kubeApiFacade.getNodeInformer().getIndexer().list();
+        List<V1Pod> pods = kubeApiFacade.getPodInformer().getIndexer().list();
+        List<Task> tasks = v3JobOperations.getTasks();
+
+        Map<String, V1Node> nodesById = nodes.stream().collect(Collectors.toMap(
+                node -> node.getMetadata().getName(),
+                Function.identity()
+        ));
+        Map<String, Task> currentTasks = tasks.stream().collect(Collectors.toMap(Task::getId, Function.identity()));
+        Set<String> currentPodNames = pods.stream().map(p -> p.getMetadata().getName()).collect(Collectors.toSet());
+
+        transitionOrphanedTasks(currentTasks, currentPodNames, nodesById);
+    }
+
+    /**
+     * Transition orphaned tasks to Finished that don't exist in Kubernetes.
+     */
+    private void transitionOrphanedTasks(Map<String, Task> currentTasks, Set<String> currentPodNames, Map<String, V1Node> nodes) {
+        List<Task> tasksNotInApiServer = currentTasks.values().stream()
+                .filter(t -> shouldTaskBeInApiServer(t) && !currentPodNames.contains(t.getId()))
+                .collect(Collectors.toList());
+
+        Map<OrphanedKind, List<Task>> orphanedTasksByKind = new HashMap<>();
+        for (Task task : tasksNotInApiServer) {
+            if (task.getStatus().getState().equals(TaskState.KillInitiated)) {
+                orphanedTasksByKind.computeIfAbsent(OrphanedKind.KILL_INITIATED, s -> new ArrayList<>()).add(task);
+            } else {
+                if (findNode(task, nodes).isPresent()) {
+                    orphanedTasksByKind.computeIfAbsent(OrphanedKind.UNKNOWN, s -> new ArrayList<>()).add(task);
+                } else {
+                    orphanedTasksByKind.computeIfAbsent(OrphanedKind.NODE_TERMINATED, s -> new ArrayList<>()).add(task);
+                }
+            }
+        }
+
+        orphanedTasksByKind.forEach((kind, tasks) -> {
+            logger.info("Attempting to transition {} orphaned tasks to finished ({}): {}", tasks.size(), kind, tasks);
+            orphanedTaskGauges.get(kind).set(tasks.size());
+
+            for (Task task : tasks) {
+
+                String reasonCode;
+                String reasonMessage;
+                switch (kind) {
+                    case KILL_INITIATED:
+                        reasonCode = TaskStatus.REASON_TASK_KILLED;
+                        reasonMessage = "Task killed";
+                        break;
+                    case NODE_TERMINATED:
+                        reasonCode = TaskStatus.REASON_TASK_LOST;
+                        reasonMessage = "Agent terminated";
+                        break;
+                    case UNKNOWN:
+                    default:
+                        reasonCode = TaskStatus.REASON_TASK_LOST;
+                        reasonMessage = "Task lost between control plane and machine";
```

Comment:
suggestion, something like: `Abandoned with unknown state due to lack of status updates from the host machine`